### ISSUE TITLE
Subscribe and Unsubscribe for non-Private Channels

### ISSUE THREAD

**Issue by robmoffat** [finos/FDC3#1165]
## Enhancement Request

The current schema for desktop agent bridging has messages:

```
PrivateChannel.onAddContextListener
PrivateChannel.onUnsubscribe
```

However, in order to reduce chatter through the desktop agent it would be good if we had these same messages for other channel types, i.e.:

```
onAddContextListener
onUnsubscribe
```

That way, the desktop agent or bridge would be able to filter traffic to only the clients that needed to receive it.  Without this, if desktop agents 1 and 2 were bridging, but only desktop agent 1 had apps connected to app channel `xzy` then the bridge would not need to broadcast messages to desktop agent 2 on channel `xzy`.

(Raising this issue here as @kriswest and I discussed it briefly in a meeting the other week and I wanted to keep a placeholder)

**Comment by kriswest** [comment]
> That way, the desktop agent or bridge would be able to filter traffic to only the clients that needed to receive it. Without this, if desktop agents 1 and 2 were bridging, but only desktop agent 1 had apps connected to app channel xzy then the bridge would not need to broadcast messages to desktop agent 2 on channel xzy.

I don't think that's true, as desktop agent 2 needs to maintain the state of channel xzy in case someone joins it (when they should receive the current state/most recent context of types they listen for). To resolve that brings in some additional complexity on join that we decided was worse when writing the protocol.

That said, I don't see a downside to those events being available on all channels, they don't currently provide details on who added the listener, only that one was added. Don't forget the `onDisconnect` event also - although its not as useful as the other two.

**Comment by robmoffat** [comment]
Disconnect only exists on private channels, doesn't it?

> I don't think that's true, as desktop agent 2 needs to maintain the state of channel xzy in case someone joins it 

only for user channels though, right?

**Comment by kriswest** [comment]
> only for user channels though, right?

No also app channels - any channel that is open access. I think it differs for private channels as DAs are supposed to track who's involved and only send it if there is someone.

**Comment by kriswest** [comment]
The additional tracking thats used for private channels is described here: https://fdc3.finos.org/docs/next/agent-bridging/spec#privatechannels

All messages broadcast to the channel are sent to the DA that created it, rather than all DAs as is the case with both other channel types. The creating DA then repeats on to only those other DAs with current subscribers.  

Whereas for (all) other channels, a copy of their state is received on connection to the bridge and then maintained with each broadcast. The alternative is to introduce a new flow to retrieve and synchronize state on retrieving a channel object, joining a channel etc., which would add a fair bit of additional complexity (particularly where the state between agents conflicts - which will happen if apps join and leave channels) and introduce delays on join.

**Comment by kriswest** [comment]
CLosed after discussion at #1166
